File: src/main/java/com/github/kindrat/cassandra/client/util/StreamUtils.java
package com.github.kindrat.cassandra.client.util;

import lombok.experimental.UtilityClass;

import java.util.Collection;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@UtilityClass
public class StreamUtils {

    public static <K, E> Map<K, E> toMap(Collection<E> collection, Function<E, K> keyMapper) {
        return toMap(collection, keyMapper, Function.identity());
    }

    public static <K, V, E> Map<K, V> toMap(Collection<E> collection,
                                            Function<E, K> keyMapper,
                                            Function<E, V> valueMapper) {
        if (collection == null) {
            return null;
        }
        return collection.stream().collect(Collectors.toMap(keyMapper, valueMapper));
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/util/StreamUtils.kt
package com.github.kindrat.cassandra.client.util

import java.util.function.Function
import java.util.stream.Collectors

object StreamUtils {
    fun <K, E> toMap(collection: Collection<E>?, keyMapper: Function<E, K>?): Map<K, E>? {
        return toMap(collection, keyMapper, Function.identity())
    }

    fun <K, V, E> toMap(collection: Collection<E>?, keyMapper: Function<E, K>?, valueMapper: Function<E, V>?): Map<K, V>? {
        return collection?.stream()?.collect(Collectors.toMap(keyMapper, valueMapper))
    }
}